File: server/service/liftRecord.go
package service

import (
	"panta/global"
	"panta/model"
	"panta/model/request"
	"panta/model/response"
	"panta/utils/enum"
	"time"
)

// @title    CreateLiftRecord
// @description   create a LiftRecord
// @param     liftRecord               model.LiftRecord
// @auth                     （2020/04/05  20:22）
// @return    err             error

func CreateLiftRecord(liftRecord model.LiftRecord) (err error) {
	liftRecord.Progress = 1
	err = global.PantaDb.Create(&liftRecord).Error
	return err
}

// @title    DeleteLiftRecord
// @description   delete a LiftRecord
// @auth                     （2020/04/05  20:22）
// @param     liftRecord               model.LiftRecord
// @return                    error

func DeleteLiftRecord(liftRecord model.LiftRecord) (err error) {
	err = global.PantaDb.Delete(liftRecord).Error
	return err
}

// @title    DeleteLiftRecordByIds
// @description   delete LiftRecords
// @auth                     （2020/04/05  20:22）
// @param     liftRecord               model.LiftRecord
// @return                    error

func DeleteLiftRecordByIds(ids request.IdsReq) (err error) {
	err = global.PantaDb.Delete(&[]model.LiftRecord{}, "id in (?)", ids.Ids).Error
	return err
}

// @title    UpdateLiftRecord
// @description   update a LiftRecord
// @param     liftRecord          *model.LiftRecord
// @auth                     （2020/04/05  20:22）
// @return                    error

func UpdateLiftRecord(params *request.LiftRecordUpdate) (err error) {
	updateMap := make(map[string]interface{})
	switch params.Progress {
	case int(enum.RecordCreated):
		updateMap["StartTime"] = time.Now()
		updateMap["WorkerId"] = params.WorkerId
		break
	case int(enum.RecordStarted):
		updateMap["Content"] = params.Content
		updateMap["EndTime"] = time.Now()
		break
	case int(enum.RecordEnded):
		updateMap["RecorderId"] = params.RecorderId
		break
	default:
		global.PantaLog.Warning("unknown lift record progress, ", params.Progress)
	}
	updateMap["Progress"] = params.Progress + 1
	err = global.PantaDb.Model(model.LiftRecord{}).Where("ID = ?", params.ID).Updates(updateMap).Error
	return err
}

// @title    GetLiftRecord
// @description   get the info of a LiftRecord
// @auth                     （2020/04/05  20:22）
// @param     id              uint
// @return                    error
// @return    LiftRecord        LiftRecord

func GetLiftRecord(id uint) (err error, liftRecord model.LiftRecord) {
	err = global.PantaDb.Where("id = ?", id).Preload("Lift").Preload("Category").Preload("Medias").
		Preload("Worker").Preload("Recorder").First(&liftRecord).Error
	return
}

// @title    GetLiftRecordInfoList
// @description   get LiftRecord list by pagination, 分页获取LiftRecord
// @auth                     （2020/04/05  20:22）
// @param     info            PageInfo
// @return                    error

func GetLiftRecordInfoList(info request.LiftRecordSearch) (err error, list interface{}, total int64) {
	limit := info.PageSize
	offset := info.PageSize * (info.Page - 1)
	// 创建db
	db := global.PantaDb.Model(&model.LiftRecord{})
	var liftRecords []model.LiftRecord
	if info.Key != "" {
		var res []response.LiftRecordCountByType
		if info.Key == "type" {
			db = db.Count(&total).Select("category_id as type,Count(*) as total").Group("category_id").
				Order("total desc, category_id desc")
			if info.Threshold != 0 {
				db = db.Having("total > ?", info.Threshold)
			}
			if info.Limit != 0 {
				db = db.Limit(info.Limit)
			}
			err = db.Scan(&res).Error
		}
		return err, res, total
	} else {
		err = db.Count(&total).Error
		err = db.Limit(limit).Offset(offset).Preload("Lift").Preload("Category").Preload("Medias").
			Preload("Worker").Preload("Recorder").Find(&liftRecords).Error
		return err, liftRecords, total
	}
}

File: web/src/api/liftRecord.js
import service from '@/utils/request'

// @Tags LiftRecord
// @Summary 创建LiftRecord
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body model.LiftRecord true "创建LiftRecord"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /liftRecord/createLiftRecord [post]
export const createLiftRecord = (data) => {
     return service({
         url: "/liftRecord/createLiftRecord",
         method: 'post',
         data
     })
 }

 export const fillLiftRecord = (data) => {
  return service({
      url: "/liftRecord/fillLiftRecord",
      method: 'post',
      data
  })
}

export const reviewLiftRecord = (data) => {
  return service({
      url: "/liftRecord/reviewLiftRecord",
      method: 'post',
      data
  })
}


// @Tags LiftRecord
// @Summary 删除LiftRecord
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body model.LiftRecord true "删除LiftRecord"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"删除成功"}"
// @Router /liftRecord/deleteLiftRecord [delete]
 export const deleteLiftRecord = (data) => {
     return service({
         url: "/liftRecord/deleteLiftRecord",
         method: 'delete',
         data
     })
 }

// @Tags LiftRecord
// @Summary 删除LiftRecord
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body request.IdsReq true "批量删除LiftRecord"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"删除成功"}"
// @Router /liftRecord/deleteLiftRecord [delete]
 export const deleteLiftRecordByIds = (data) => {
     return service({
         url: "/liftRecord/deleteLiftRecordByIds",
         method: 'delete',
         data
     })
 }

// @Tags LiftRecord
// @Summary 更新LiftRecord
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body model.LiftRecord true "更新LiftRecord"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"更新成功"}"
// @Router /liftRecord/updateLiftRecord [put]
 export const updateLiftRecord = (data) => {
     return service({
         url: "/liftRecord/updateLiftRecord",
         method: 'put',
         data
     })
 }


// @Tags LiftRecord
// @Summary 用id查询LiftRecord
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body model.LiftRecord true "用id查询LiftRecord"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"查询成功"}"
// @Router /liftRecord/findLiftRecord [get]
 export const findLiftRecord = (params) => {
     return service({
         url: "/liftRecord/findLiftRecord",
         method: 'get',
         params
     })
 }


// @Tags LiftRecord
// @Summary 分页获取LiftRecord列表
// @Security ApiKeyAuth
// @accept application/json
// @Produce application/json
// @Param data body request.PageInfo true "分页获取LiftRecord列表"
// @Success 200 {string} string "{"success":true,"data":{},"msg":"获取成功"}"
// @Router /liftRecord/getLiftRecordList [get]
 export const getLiftRecordList = (params) => {
     return service({
         url: "/liftRecord/getLiftRecordList",
         method: 'get',
         params
     })
 }